### ISSUE TITLE
Unit UI: add the possibility to show a UMC stat

### ISSUE THREAD

**Issue by spixi** [wesnoth/wesnoth#8888]
### Describe the desired feature

Currently, units only support two stats which are represented using bars: HP and XP. These two stats are hardcoded in ```scr/units/drawer.cpp```. UMC may make use of additional stats like mana or charge. The desired feature would be an additional user-made stat which has a current value, maximum value, a color, a name and a tooltip. The stat should be shown on the side bar and in the recall detail page next to the HP stat and as a bar overlay on the unit.

The following stat should be able to be set in [unit_type] and [unit], it should be also possible to change them with [modify_unit]
- user_stat (integer)
- max_user_stat (integer)
- user_stat_color (color string)
- user_stat_name (string)
- user_stat_tooltip (pango markup string)

This mockup shows like such a user-made stat could look like:
![image](https://github.com/wesnoth/wesnoth/assets/3192596/39d1b859-3ef6-4b60-8017-e14b2ea3c056)

**Comment by CelticMinstrel** [comment]
I understand the intent, and don't really have a problem with it, but this suggestion as described seems pretty bad to me.

**Comment by spixi** [comment]
> I understand the intent, and don't really have a problem with it, but this suggestion as described seems pretty bad to me.

What specifically do you seems pretty bad to you in this proposal? This is the most generic approach and does not require an additional meta object like [user_made_stat]. It also allows different UMC stats for different units and complex unit-specific tooltips. For example, different units may use different spells or ships have different ressources and units on board.

**Comment by CelticMinstrel** [comment]
> the most generic approach

It may or may not be a good thing, but I'm pretty sure a more generic approach could be conceived. I'm not saying the problem is that it's not generic enough though.

> does not require an additional meta object like [user_stat]

I don't know why this counts as a plus. (But I'm not saying it counts as a minus either.)

----

More to the point, what advantage does this proposal actually offer over what you can do _right now_? From what I can tell, the only advantage would be that you don't need a custom theme. I'm pretty sure everything you're describing here is possible right now with a custom theme though. Maybe it could be tricky to actually place the new stat where you want it in the theme, but if so, that's entirely a shortcoming in the theming system.

**Comment by spixi** [comment]
> More to the point, what advantage does this proposal actually offer over what you can do _right now_? From what I can tell, the only advantage would be that you don't need a custom theme. I'm pretty sure everything you're describing here is possible right now with a custom theme though. Maybe it could be tricky to actually place the new stat where you want it in the theme, but if so, that's entirely a shortcoming in the theming system.

Yes, exactly that is the point. You can already have custom menus and dialogs, where you can also show user-defined variables. It is also possible to overlay units with images, but not so trivial. You constantly have to remove and add overlays and use IPFs like CROP to achieve a similar appearance like in the mockup. However, there is no consistent and easy to use way to show a user-defined stat in several places, where a player would expect it.

**Comment by CelticMinstrel** [comment]
Sure, but I think the missing piece you need to make it work is not a `user_stat` in unit/unit_type but rather something like that in `wesnoth.interface.game_display`. I suppose there's _some_ merit in linking it to the unit type rather than having to set it up on each individual unit though… but not sure I like this exact proposal as stated. Still… I'd say the theme item should be the focus here, not whatever the source data might look like in the unit.

**Comment by gfgtdf** [comment]
> I'm pretty sure everything you're describing here is possible right now with a custom theme though

The picture in the first post shows the recruitment dialog too and you cannot modify that with a theme. This is imo the main problem with the "theme" solution for those kinds of problems and why people for example use dummy traits to show custom information I stead of theme items (because they want to show this information in the recruitment dialog too)

**Comment by CelticMinstrel** [comment]
> The picture in the first post shows the recruitment dialog too and you cannot modify that with a theme.

You do have a good point there, admittedly.

I'm still not sure that the described solution is the best though.

**Comment by babaissarkar** [comment]
@CelticMinstrel 

> I'm still not sure that the described solution is the best though.
If you can explain how this can be made better, I'm willing to implement this.
